src/manager/topdown.rs

```diff
@@ -0,0 +1,247 @@
+                            .ipc_validator_has_voted_topdown(
+                                // FIXME: Do not use the default, use the one configured
+                                // for the subnet
+                                &Address::from_str(GATEWAY_ACTOR_ADDRESS)?,
+                                submission_epoch,
+                                account,
+                            )
+                            .await
+                            .map_err(|e| {
+                                log::error!(
+                                    "error checking if validator has voted in subnet: {:?}",
+                                    &child.id
+                                );
+                                e
+                            })?;
+
+                        if !has_voted {
+                            // submitting the checkpoint synchronously and waiting to be committed.
+                            submit_topdown_checkpoint(
+                                submission_epoch,
+                                parent_tip_set,
+                                child_tip_set,
+                                account,
+                                child.id.clone(),
+                                &child_client,
+                                &parent_client,
+                            )
+                            .await?;
+
+                            // check if by any chance we have the opportunity to submit any outstanding checkpoint we may be
+                            // missing in case the previous one was executed successfully.
+                            // - we get the up to date head of the parent and the child.
+                            // - check the last executed checkpoint for the subnet
+                            // - And if we still have the info, submit a new checkpoint
+                            let parent_head = parent_client.chain_head().await?;
+                            let curr_epoch: ChainEpoch = ChainEpoch::try_from(parent_head.height)?;
+                            let cid_map = parent_head.cids.first().unwrap().clone();
+                            let parent_tip_set = Cid::try_from(cid_map)?;
+                            let child_head = child_client.chain_head().await?;
+                            let cid_map = child_head.cids.first().unwrap().clone();
+                            let child_tip_set = Cid::try_from(cid_map)?;
+                            let child_gw_state =
+                                child_client.ipc_read_gateway_state(child_tip_set).await?;
+                            let last_exec = child_gw_state
+                                .top_down_checkpoint_voting
+                                .last_voting_executed;
+                            let submission_epoch = last_exec + period;
+                            if curr_epoch >= submission_epoch {
+                                submit_topdown_checkpoint(
```

Comment:
If the validator is lagging the current epoch, we have to speed up the submission. Otherwise, the checkpoint would be processed too slow.

Reply:
Already fixed with the loop. This will be improved further with the refactor.